### ISSUE TITLE
Cannot build using Mac OS

### ISSUE THREAD

**Issue by luiscruz** [hubblo-org/scaphandre#46]
## Bug description

Cannot build using Mac OS.

## To Reproduce

`cargo build`

## Relevant Output

```
error[E0433]: failed to resolve: could not find `linux` in `os`
  --> /Users/luiscruz/.cargo/registry/src/github.com-1ecc6299db9ec823/procfs-0.8.1/src/process/mod.rs:62:14
   |
62 | use std::os::linux::fs::MetadataExt;
   |              ^^^^^ could not find `linux` in `os`

error[E0599]: no method named `st_uid` found for struct `Metadata` in the current scope
   --> /Users/luiscruz/.cargo/registry/src/github.com-1ecc6299db9ec823/procfs-0.8.1/src/process/mod.rs:562:23
    |
562 |             owner: md.st_uid(),
    |                       ^^^^^^ method not found in `Metadata`
    |
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope; perhaps add a `use` for it:
            `use std::os::macos::fs::MetadataExt;`

error[E0599]: no method named `st_mode` found for struct `Metadata` in the current scope
   --> /Users/luiscruz/.cargo/registry/src/github.com-1ecc6299db9ec823/procfs-0.8.1/src/process/mod.rs:763:30
    |
763 |                     mode: md.st_mode() & libc::S_IRWXU,
    |                              ^^^^^^^ method not found in `Metadata`
    |
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope; perhaps add a `use` for it:
            `use std::os::macos::fs::MetadataExt;`

error[E0308]: mismatched types
   --> /Users/luiscruz/.cargo/registry/src/github.com-1ecc6299db9ec823/procfs-0.8.1/src/process/mod.rs:182:22
    |
182 |         const READ = libc::S_IRUSR;
    |                      ^^^^^^^^^^^^^ expected `u32`, found `u16`

error[E0308]: mismatched types
   --> /Users/luiscruz/.cargo/registry/src/github.com-1ecc6299db9ec823/procfs-0.8.1/src/process/mod.rs:183:23
    |
183 |         const WRITE = libc::S_IWUSR;
    |                       ^^^^^^^^^^^^^ expected `u32`, found `u16`

error[E0308]: mismatched types
   --> /Users/luiscruz/.cargo/registry/src/github.com-1ecc6299db9ec823/procfs-0.8.1/src/process/mod.rs:184:25
    |
184 |         const EXECUTE = libc::S_IXUSR;
    |                         ^^^^^^^^^^^^^ expected `u32`, found `u16`

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0308, E0433, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `procfs`
```
## Environment

- Mac OS 10.15.7
- Kernel version: 19.6.0

## Context

I have just realised that this project is targeting linux distributions only. That kinda makes sense from a production perspective. From a dev perspective is quite limiting. Any ideas?

**Comment by bpetit** [comment]
Hi,

Thanks for the try. This was somehow expected. The powercaprapl sensor is based on a linux kernel module. We will need to create a new sensor for mac os I guess. Discussions about that have started in the ARM related issue (about apple m1) that may be redundant but just to mention.

I have an old macbook air somewhere so I may start digging on that topic but not before february/march. If anyone with a macbook (maybe better to start with x86 macbooks to tackle one topic at a time) is willing to give it a shot before, it would be awesome :)

**Comment by luiscruz** [comment]
I'd be happy to help but that might be a bit too much for me…
Still, it feels like it should be possible to build without having a Mac OS sensor ready. I can imagine that a Mac OS sensor will not be compatible with Linux, but it should still be possible to build it on a Linux machine.

**Comment by mrchrisadams** [comment]
Hi @luiscruz I looked into some of this myself over the holidays.

My notes are here for OS X, but the TLDR is that if we want RAPL data, we likely need to look for the output from apple's `powermetrics` cli command.

https://github.com/hubblo-org/scaphandre/issues/35

**Comment by mrchrisadams** [comment]
Ah, here's what you can get out. You can access the RAPL code with a command like so:

```
sudo powermetrics -i 2 -n 1 -s cpu_power --format plist
```

this runs powermetrics, pulling out only the cpu_power info, and formats it as a dialect of XML (the `--format plist`) bit, taking a sample for 2 milliseconds, and doing it once.

I can't see a way to read a stream of info through :(

**Comment by florimondmanca** [comment]
FWIW I just realized I'm oftentimes using [CoconutBattery](https://www.coconut-flavour.com/coconutbattery/), which is a macOS utility to view battery state and usage. It has a "show watt usage" thing. I'd be curious to know how it does that. It doesn't seem to be open source (hmm… should I be using it then? 😝), but the author's email is at the bottom of the landing page. (I don't know their github @, else we could have pinged them here…)

**Comment by bpetit** [comment]
Great to see there are userspace tools that get such metrics, this is a good sign !

Looking at your message @florimondmanca I just want to mention that battery based metrics, may be a bit different than rapl. Like building a macos sensor based on battery metrics wouldn't work on a mac mini server for instance. It's not necessary an issue, if we name the sensor appropriately and document it to indicate that this is only suitable for a laptop. (it reminds me of powertop, that gets metrics from battery related sensors if available and fallbacks to something else if not...)

@mrchrisadams it would be interessting to look at what libraries/bsd internals powermetrics calls. Maybe there we could get a stream/more suitable sort of data ?

**Comment by bpetit** [comment]
@luiscruz yes maybe we could build. But the powercaprapl sensor being the only available right now it wouldn't be very helpful for mas users to have a binary that gets no metrics :/

**Comment by mrchrisadams** [comment]
So, on OS X there is this:

https://github.com/beltex/dshb - a nice open source CLI exposing the similar stats as coconut battery.

Under the hood, that uses this library, in Swift:

https://github.com/beltex/SystemKit

And that in turn uses an open sourced library from apple:
https://opensource.apple.com/source/PowerManagement/

And this issue here  in system kit is about where I ran out of steam when looking at it.
https://github.com/beltex/SystemKit/issues/2

I think you'd need to find someone who really understands the powermetrics code to provide pointers, and get an idea of what it's reading to get these numbers.

I don't think the powermetrics is open, and but this post here might shed some more light:
https://www.cmdsec.com/macos-performance-monitoring-collection/

The metrics are exposed, because iStat pro shows me stuff like this on my mac:
![Screenshot 2021-01-07 at 17 04 09](https://user-images.githubusercontent.com/17906/103914748-99652e80-510a-11eb-82d8-a81f351e4f09.png)


But I don't have experience working this close to the metal with OS X.

**Comment by bpetit** [comment]
That may be a good news for developing scaphandre support on the m1:

![2021-02-02_09-46](https://user-images.githubusercontent.com/906428/106574871-5775af80-6533-11eb-8583-049037dc9030.png)

**Comment by bpetit** [comment]
> The `procfs` crate is designed specifically for Linux. From the brief discussion in [eminence/procfs#42](https://github.com/eminence/procfs/issues/42) it looks like an alternative approach is needed for macOS, probably using [conditional compilation](https://doc.rust-lang.org/reference/conditional-compilation.html) to bring in a separate platform specific library on macOS and [Windows](https://github.com/hubblo-org/scaphandre/issues/74).

Absolutely. We need to rewrite a part of the sensors logic. I'll should start this refactoring next week as I plan to work on the windows support (#74).

**Comment by Vic063** [comment]
Hello,

I just looked into _Intel Power Gadgets_ which uses a kernel driver to read/write from/into MSRs.
Sadly, I do not own an Apple computer so I only tested on a VM running under qemu/KVM.

Obviously, required MSRs are not available but I have a list of the used MSRs (which are mainly the same used by the Windows version):
[244872.306699] kvm [927770]: ignored rdmsr: 0x606 data 0x0
[244872.306699] kvm [927770]: ignored rdmsr: 0x610 data 0x0
[244872.306699] kvm [927770]: ignored rdmsr: 0x611 data 0x0
[244872.306699] kvm [927770]: ignored rdmsr: 0x614 data 0x0
[244872.306710] kvm [927770]: ignored rdmsr: 0x38f data 0x0
[244872.306714] kvm [927770]: ignored rdmsr: 0x38d data 0x0
[244872.306724] kvm [927770]: ignored rdmsr: 0x38f data 0x0

However, I use to run a modified Linux kernel which implements these MSRs so it could be possible to inject (a/k/a write) values into these MSRs and make them available to Scaphandre inside the VM (I already did that for Windows VMs).

I also looked into the solution purposed by @mrchrisadams but _syspower_ does not return any value on my Mac OS VM. I did not dig more than that.

Now I'm thinking about writing an OS X driver with the capability to read MSRs values, like the Windows driver. As the Scaphandre engine is being rewritten to support multiple OSes, the same method would be applied to read from OS X or Windows NT, right @bpetit?

**Comment by bpetit** [comment]
Hi !

Yes @Vic063 I think we could use the same methodology in the Scaphandre codebase to plug to a dedicated MSR based driver. Even if we find a less invasive way to monitor power consumption on MacOS at some point, it may be included in a new sensor, in addition to this one anyway. So your help would be very helpful here 😁  (but we also have to release the windows driver codebase soon :P).

I also add to this topic this recent very interesting blog post : https://www.rdegges.com/2022/how-to-calculate-the-energy-consumption-of-a-mac/

It seems the author uses battery data, which is something we could have used for Linux technically, but didn't use because it didn't fit a server use case. However, on this MacOs topic, I have the feeling that most of the use cases are for laptops (if anyone has a mac mini or a mac server raise a hand ! 😂). So maybe, if getting RAPL data on macos is too time consuming, we could use the data from battery consumption as a first iteration ? WDYT ? (I don't say we should do it, but having a sensor based on those data seems interessting too, in this context, especially for MacOS and M1 CPU that are ARM CPU and don't provide RAPL for example...).

**Comment by rdegges** [comment]
Happy to test stuff out for ya -- the approach I outlined in that article is using Apple's `ioreg` CLI utility which inspects hardware devices and returns information. They have a low-level API within their "battery" metrics that allows you to read the current amount of watts being pulled from a power adapter -- which is what I used to measure total consumption over time.

In my experience, this is the best way to do it if you want to get an accurate reading (I verified it was accurate using hardware monitors) for the OVERALL energy consumption of the device. The downside (I think?) for your usecase is that it returns the number of watts overall, it isn't looking at how many volts particular components are using at any given moment (CPU vs GPU etc.).

**Comment by davidmytton** [comment]
Firefox 104 introduced a new power profiler for [Windows](https://hg.mozilla.org/mozilla-central/rev/185ab81e5f51) (Windows 11 or Windows 10 with an [EMI](https://docs.microsoft.com/en-us/windows-hardware/drivers/powermeter/energy-meter-interface)) and [Mac](https://hg.mozilla.org/mozilla-central/rev/0839b2f7ae02) (Apple Silicon only). This provides output in watts in the profiler UI, but the internals are [measuring in picowatt-hour](https://hg.mozilla.org/mozilla-central/file/tip/tools/profiler/core/PowerCounters-mac.cpp#l35). On Mac it's using the [task_info API](https://github.com/apple/darwin-xnu/blob/8f02f2a044b9bb1ad951987ef5bab20ec9486310/osfmk/mach/task_info.h#L464).

Discussed in https://www.green-coding.org/blog/firefox-104-energy-measurements/

**Comment by boynazarovnurbekjon62-netizen** [comment]
<img width="1254" height="1254" alt="Image" src="https://github.com/user-attachments/assets/0b166898-9214-430d-8475-a21be992a888" /> ushbu ilovani yaratishimdan maqsad odamlarga foydam tegishidan xursandman

**Comment by boynazarovnurbekjon62-netizen** [comment]
Markdownd 

<img width="1254" height="1254" alt="Image" src="https://github.com/user-attachments/assets/1a6b3f0f-f9ee-4f0f-964c-07337780038f" /> ilovamni yarataman

**Comment by boynazarovnurbekjon62-netizen** [comment]
Markdownd 

<img width="1254" height="1254" alt="Image" src="https://github.com/user-attachments/assets/22d58a09-172a-493f-a054-00184789d595" />

**Comment by boynazarovnurbekjon62-netizen** [comment]
> Markdownd
> 
> <img alt="Rasm" width="1254" height="1254" src="https://private-user-images.githubusercontent.com/332205905/656033184-22d58a09-172a-493f-a054-00184789d595.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3OTAwMjE2NjcsIm5iZiI6MTc5MDAyMTM2NywicGF0aCI6Ii8zMzIyMDU5MDUvNjU2MDMzMTg0LTIyZDU4YTA5LTE3MmEtNDkzZi1hMDU0LTAwMTg0Nzg5ZDU5NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwOTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDkyMVQyMDA5MjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iODdjMDlkYzEyOTg1YmQ5ZTNlYTAzYzNiMTFmNDUxZGIxODQ5YmUxMjJlOTFjMTdjZTQ3MzMxZGQ0NTRhODg5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ._Li-obgdz5ZFbNxo_VPkxAmo1mzkRZtyrrTSa0CociE">
